yorkie/backend/sync/etcd/client.go

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2021 The Yorkie Authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package etcd
+
+import (
+	"context"
+	"time"
+
+	"go.etcd.io/etcd/clientv3"
+	"go.etcd.io/etcd/clientv3/concurrency"
+
+	"github.com/yorkie-team/yorkie/pkg/log"
+	"github.com/yorkie-team/yorkie/yorkie/backend/sync"
+)
+
+// DefaultLockLeaseTimeSec is the default lease time of lock.
+const DefaultLockLeaseTimeSec = 30
+
+// Config is the configuration for creating a Client instance.
+type Config struct {
+	Endpoints      []string      `json:"Endpoints"`
+	DialTimeoutSec time.Duration `json:"DialTimeoutSec"`
+	Username       string        `json:"Username"`
+	Password       string        `json:"Password"`
+
+	LockLeaseTimeSec int `json:"LockLeaseTimeSec"`
+}
+
+// Client is a client that connects to ETCD.
+type Client struct {
+	config *Config
+	client *clientv3.Client
+}
+
+// newClient creates a new instance of Client.
+func newClient(conf *Config) *Client {
+	if conf.LockLeaseTimeSec == 0 {
+		conf.LockLeaseTimeSec = DefaultLockLeaseTimeSec
+	}
+
+	return &Client{
+		config: conf,
+	}
+}
+
+// Dial creates a new instance of Client and dials the given ETCD.
+func Dial(conf *Config) (*Client, error) {
+	c := newClient(conf)
+
+	if err := c.Dial(); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Dial dials the given ETCD.
+func (c *Client) Dial() error {
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   c.config.Endpoints,
+		DialTimeout: c.config.DialTimeoutSec * time.Second,
+		Username:    c.config.Username,
+		Password:    c.config.Password,
+	})
+	if err != nil {
+		log.Logger.Error(err)
+		return err
+	}
+	log.Logger.Infof("etcd connected, URI: %s", c.config.Endpoints)
+
+	c.client = cli
+	return nil
+}
```

Comment:
How about checking the actual connection?
Currently, it is not possible to check the connection status to the ETCD endpoint.

### If not intended:

Actually, I tested `TestETCD` of `etcd_test.go` without running `ETCD` docker. And even if it couldn't actually be connected, it was confirmed that the test passed.
Even if you run `make test` in the absence of `ETCD` docker, the test will not proceed somewhere and will be in a standby state. (**Where the problem occurs is another matter.**)

While I analyzing this problem, I found that when I wrote the code below, I could check the connection.
[Reference]: https://github.com/etcd-io/etcd/issues/9877

In the issue I referenced, `[]grpc.DialOption{grpc.WithBlock()}` was mentioned, but in our code it did not work as intended. If we look into the reasons together, it seems that simpler code can solve this problem.
```go
func (c *Client) Dial() error {
	cli, err := clientv3.New(clientv3.Config{
		Endpoints:   c.config.Endpoints,
		DialTimeout: c.config.DialTimeoutSec * time.Second,
		Username:    c.config.Username,
		Password:    c.config.Password,
	})
	if err != nil {
		log.Logger.Error(err)
		return err
	}

	timeoutCtx, cancel := context.WithTimeout(context.Background(), c.config.DialTimeoutSec*time.Second)
	defer cancel()

	_, err = cli.Status(timeoutCtx, c.config.Endpoints[0])
	if err != nil {
		log.Logger.Error(err)
		return err
	}
        ...
	return nil
}
```

Reply:
Thank you for telling me. I modified it by referring to the PR below.
https://github.com/kubernetes/kubernetes/pull/81435